From 410e41862f5e4d09cf75ad723dd5ad2c1c711823 Mon Sep 17 00:00:00 2001
From: Ran Shalit <rshalit@habana.ai>
Date: Wed, 08 Jul 2020 14:42:47 +0300
Subject: [PATCH] [SW-15860] gaudi: security: create otp blob

Create fit file and otp blob containing keys
and header, and new buildroot target,
create-otp-image.

Change-Id: I2b99ba9da3dc3d9a20cc13dd8b52c1e49c11fb25
Signed-off-by: Ran Shalit <rshalit@habana.ai>
---

diff --git a/board/habanalabs/common/key/u-boot/create-otp-image.sh b/board/habanalabs/common/key/u-boot/create-otp-image.sh
new file mode 100755
index 0000000..95e1b0b
--- /dev/null
+++ b/board/habanalabs/common/key/u-boot/create-otp-image.sh
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
+
+OTP_MAGIC_NUMBER=0x444F4C48
+OTP_HEADER_VERSION=0x1
+OTP_SHA_CRYPTO_TYPE=0x1
+OTP_KEYS_COUNT=0x5
+OTP_RESERVED=0x0
+
+BOARD_BASE=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}
+# MKIMAGE_FILE=${BR2_EXTERNAL}/output/${HL_DEV}/host/bin/mkimage
+KEY_DIR=${BR2_EXTERNAL}/board/habanalabs/common/key/u-boot
+KEY_FILE1_NAME=hl_rsa3072.key
+KEY_FILE2_NAME=hl_rsa3072.key
+KEY_FILE3_NAME=hl_rsa3072.key
+KEY_FILE4_NAME=hl_rsa3072.key
+KEY_FILE5_NAME=hl_rsa3072.key
+OTP_IMAGE_NAME=otp.bin
+
+OTP_IMAGE=${KEY_DIR}/${OTP_IMAGE_NAME}
+
+#TBD check if rev_bytes_stdin required !
+echo "======================"
+echo $OTP_IMAGE
+#write u32 magic number
+rm $OTP_IMAGE
+printf "%08x" $OTP_MAGIC_NUMBER 	| xxd -r -p - $OTP_IMAGE
+#write u8 header version
+printf "%02x" $OTP_HEADER_VERSION	| xxd -r -p -s 4 - $OTP_IMAGE
+printf "%02x" $OTP_SHA_CRYPTO_TYPE 	| xxd -r -p -s 5 - $OTP_IMAGE
+printf "%02x" $OTP_KEYS_COUNT 		| xxd -r -p -s 6 - $OTP_IMAGE
+printf "%02x" $OTP_RESERVED 		| xxd -r -p -s 7 - $OTP_IMAGE
+
+#write keys #1 to #5 into file
+
+sha256sum -b ${KEY_DIR}/${KEY_FILE1_NAME} | cut -d' ' -f1 | xxd -r -p -s 8 - $OTP_IMAGE	
+sha256sum -b ${KEY_DIR}/${KEY_FILE2_NAME} | cut -d' ' -f1 | xxd -r -p -s 40 - $OTP_IMAGE	
+sha256sum -b ${KEY_DIR}/${KEY_FILE3_NAME} | cut -d' ' -f1 | xxd -r -p -s 72 - $OTP_IMAGE		
+sha256sum -b ${KEY_DIR}/${KEY_FILE4_NAME} | cut -d' ' -f1 | xxd -r -p -s 104 - $OTP_IMAGE		
+sha256sum -b ${KEY_DIR}/${KEY_FILE5_NAME} | cut -d' ' -f1 | xxd -r -p -s 136 - $OTP_IMAGE		
+
diff --git a/public b/public
new file mode 100644
index 0000000..55f7b7f
--- /dev/null
+++ b/public
@@ -0,0 +1,30 @@
+-----BEGIN PGP PUBLIC KEY BLOCK-----
+Version: GnuPG v2.0.22 (GNU/Linux)
+
+mQENBFxSC0YBCADCIeuaXZ1Cixe5aqeZ614F7OnQslN7RgeKh0nBeNdV9x86adc8
+g9uZUpzwdD2qCe+DUj6H9RADnXJoz8nRR4r28iC9Tw2hyy6Rzkgh9pXjSW47yDX0
+eM0PB8L1zOVDNAtth9+JRk5phXDsfoxE/XC1IbhIoedJECxUXDeuG81u7qs3X1Pl
+91QBRaGznAjkHxstdag0ieDDCYrRZgDBJ9hQGTlCO0VH35w1zobSJsxljgnaSZbA
+EmOTPDqbj6aXrCdGC98mjeVFca3ltUlk+MQ/NcSYxxFvxYOhflkInCFyeMNCoDH5
+edS0B1HFBe72AoydtFrWyh5mwrRpDsQiqz5JABEBAAG0KkhhYmFuYS5haSAoQXJ0
+aWZhY3RvcnkpIDxkZXZvcHNAaGFiYW5hLmFpPokBOQQTAQIAIwUCXFILRgIbAwcL
+CQgHAwIBBhUIAgkKCwQWAgMBAh4BAheAAAoJEB5nbOg2ot5lA1AIAKZpT6OY0zzl
+oZeIVpXb6Usl31PE+VEyocFlr4T4RcCWGce1Ib+MCk9iYGOzGyJO6y+icSXKiWiD
+VPoU281ge68eMVfIdY3KEjfUeWD+pC+gM/CTVUw2CqDTY1LmvMfit8/u40CjR06j
+voxldMY8U/yvE3CAozQiRUlkLbi0Qkaxgy22ue7EevT3gRxMyXaTCJrrTQH7ZUC6
+0qV6aLau5lmzrScWBdbJFx7GiQRP/Pi44G5NvpWDTKmUeDPfPrKcoeag0GiTrzze
+lI1qkutjdqFfsE/87fQ8tedy0EV2GC0qipCIScNPMvwjU8NJAXE8umfr1Z+bz+2q
+QrUjh4xTNbS5AQ0EXFILRgEIALA0f38xsZ5K9VThZ4jhlAHmzP06KdGDuZKeAO21
+fgK3QxPt2ahsb459Q3Wyr1vztm104JmACXICheAhLWGkGhe/PjxPcwKAjY1YiL1r
+jff7BBsZFe6j8FiN51qrAbeBvRS9XBwtr0ZYrn3RwKu0s1fl8yJm0szDTBNpS4TN
+AL4/9+CU2ElFAt2dLfUlCYmMVLOeNIrSjqIRQ5lf9Y0rlVEgEH4ificlkd9DTXNK
+Jojtgp5TrYN/h9pvod8OB05ZNUp1jlvH6L1wlVmEhHqpZbjBKsg9EhkYCUA9woqz
++xLJvnTsWw+ErvFATXo8EWVL0IsaoW5vAu5+TlAYTX5h9J0AEQEAAYkBHwQYAQIA
+CQUCXFILRgIbDAAKCRAeZ2zoNqLeZaHqB/94g4fCwaAPT5HtFojMnTzemc+hPiB3
+FxorlBxTlHiJVo3hNeiZpytPx8XTlIaF1xzOZClHZ+2SMs87RmIVqf9osvUeRUvS
+/VzywMoC51FLIwSDUtMLEZh/QGNxKxBspA5vSlj1Vfq3HntiynGPG6BkULEXx9E9
+LZAyXv9YF/xDoWBCh8CHk0LISlLsiEcLsglbUSKGKuErtZxnHzhSxw7TparnAryW
+S4e1ExV4KjaNhKVJ59IqVa1rfhFXfGXL4kAd3aMwmv5SJo/YwytSvXXUa4iix5sd
+O2VR/AflyPJZGSYRRY6vqmNqlQfwL2SeHcVdcRGq4TEdut344MBpM9hD
+=5Ds1
+-----END PGP PUBLIC KEY BLOCK-----
diff --git a/scripts/make/gaudi/Makefile b/scripts/make/gaudi/Makefile
index 9e8e710..3380c32 100644
--- a/scripts/make/gaudi/Makefile
+++ b/scripts/make/gaudi/Makefile
@@ -69,4 +69,8 @@
 .PHONY: gaudi-asic-robot
 gaudi-asic-robot: gaudi-asic-prepare
 	$(MAKE) common-asic-tools-install HL_INSTALL_LAB_IP=$(HL_INSTALL_LAB_IP)
-	$(MAKE) common-asic-robot
\ No newline at end of file
+	$(MAKE) common-asic-robot
+	
+.PHONY: gaudi-otp-create
+gaudi-otp-create:
+	. ${BR2_EXTERNAL}/board/habanalabs/common/key/u-boot/create-otp-image.sh  
\ No newline at end of file
